### PR TITLE
Adding Support for Container APIs on Blob Endpoint

### PR DESCRIPTION
**Pull request by anujmodi2021**
Implemented following APIs on Blob Endpoint
1. Create Container
2. Delete Container
3. Set Container Metadata
4. Get Container Metadata

Get Container Properties was already implemented.

### CODE REVIEW

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/DelegationSASGenerator.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/DelegationSASGenerator.java:89]: Spacing issue

**Comment by anmolanmol1234** [comment]
Rest looks good, add a fallback mechanism for dependent API's as well.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: perfinfo registering success twice.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: common code in getFilesystemProperties as well:
```
  final AbfsRestOperation op = client.getContainerMetadata(tracingContext);
 final Hashtable<String, String> metadata = parseResponseHeadersToHashTable(op.getResult());
```

should we make seperate method and use that both places.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (saxenapranav) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: common code with getFileSystemProperties.

**Review by saxenapranav — CHANGES_REQUESTED** [review]

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (snvijaya) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: +1

**Review by snvijaya — APPROVED** [review]
Minor comment already raised by Pranav. Please proceed with checkin once Pranav approves.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/DelegationSASGenerator.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/DelegationSASGenerator.java:89]: There was a spacing issue earlier, I fixed it in this change.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: You mean setFileSystemProperties right??
Only two lines of code are getting repeated and both of them are function calls themselves...
I think it would be redundant to have another method that we will still need to call from two places.

What do you think??
We will only be saving one line repetition.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: Same as setFileSystemProperties.
We will still end up with one line repeated..

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:963]: Why would we not consider the useBlobEndpoint variable while making the decision here ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java]: perfInfo.registerResult(op.getResult()).registerSuccess(true); can be changed to a single line

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:784]: This should also change to the function that @saxenapranav  added newly right ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: URL can be made final

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsClient.java]: same final URL

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:963]: Because deleteFileSystem is not a HDFS API. This function is never called in any flow that comes from AzureBlobFileSystem. This is used only in test classes.

So, I thought it does not make sense to have fallback for this method as it will never be used.
Same is the case for other file system APIs: getFileSystemProperties() and setFileSystemProperties()

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:784]: Yes, waiting for his PR to merge.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:481]: The perfInfo line should come at 486 ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:475]: Same here at 477 ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:481]: Taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java:475]: Taken

**Review by anmolanmol1234 — APPROVED** [review]

**Review by saxenapranav — APPROVED** [review]
LGTM! Thanks for taking comments :).

**Pull request merged**
